src/Components/Components/src/RenderTree/EventHandlerDelegateWithClosuresComparer.cs

```diff
@@ -0,0 +1,130 @@
+        {
+            // if this type is not compiler generated, it is not a closure, and the previous tests where
+            // enough to conclude that these are not equal. We just cache a delegate to a function that returns false.
+            if (!type.CustomAttributes.Any(x => x.AttributeType == typeof(CompilerGeneratedAttribute)))
+            {
+                return (objA, objB) => false;
+            }
+
+            // build an array of lambdas that will get each public field value on the closure.
+            var fieldGetters = type.GetFields(
+                BindingFlags.Public |
+                BindingFlags.Instance |
+                BindingFlags.DeclaredOnly).Select(fieldInfo =>
+            {
+                var sourceParam = Expression.Parameter(typeof(object));
+                Expression returnExpression = Expression.Field(Expression.Convert(sourceParam, fieldInfo.DeclaringType!), fieldInfo);
+                if (fieldInfo.FieldType.IsValueType)
+                {
+                    // box if neccessary.
+                    returnExpression = Expression.Convert(returnExpression, typeof(object));
+                }
+                var lambda = Expression.Lambda(returnExpression, sourceParam);
+                return (Func<object?, object?>)lambda.Compile();
+            }).ToArray();
+
+            // the actual comparison function. It re-uses the array of public field 'getters' that we have built up
+            // before.
+            return (objA, objB) =>
+            {
+                // We have two instances of the same compiler generated class. Let's compare all public fields.
+                foreach (var fieldGetter in fieldGetters)
+                {
+                    var leftValue = fieldGetter(objA);
+                    var rightValue = fieldGetter(objB);
+
+                    // We might have recursive callbacks or delegates.
+                    // For example EventCallbackFactoryBinderExtensions.CreateBinderCore<T> wraps a delegate in a delegate
+                    // so this is not hypothetical.
+                    if (leftValue is EventCallback leftEventCallback && rightValue is EventCallback rightEventCallBack)
+                    {
+                        if (!EventCallBackEquals(ref leftEventCallback, ref rightEventCallBack))
+                        {
+                            return false;
+                        }
+                    }
+                    else if (leftValue is MulticastDelegate leftDelegate && rightValue is MulticastDelegate rightDelegate)
+                    {
+                        if (!DelegateEquals(leftDelegate, rightDelegate))
+                        {
+                            return false;
+                        }
+                    }
+                    else if (!Equals(leftValue, rightValue))
```

Comment:
Are you sure we should be using `Equals` like this? This means the behavior would vary depending on how `Equals` is overridden on the instance types. I would have expected something like `Object.ReferenceEquals` instead.

What's your reasoning about the kinds of equality we want to recognize here?

Reply:
Well, who am I to judge equality? If for some reason anyone decided to override equality for a type, should I ignore that when an instance of that type is captured as free variable. That would be a pretty big decision and would need a big note in the docs. Also, the original value the attribute is pointing to, is compared as Equals() as well, so it makes sense to continue that here.